packages/optimizer/lib/transformers/AmpScriptCsp.js

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const {calculateHash} = require('@ampproject/toolbox-script-csp');
+
+/**
+ * AmpScriptCsp - adds CSP for amp-script.
+ *
+ * Currently only supports inline scripts.
+ *
+ * This transformer supports the following parameter:
+ *
+ * `append`: specifies the calculated CSPs should be appended to existing ones.
```

Comment:
Do we even need this? `append=true` needs to be the default behavior] to not break existing hashes for cross-origin scripts.

Reply:
I think appending by default is a bad idea from a security perspective. The idea behind CSP is to ensure that injected scripts don't work.

If an attacker is somehow able to manipulate the pre-optimizer code and add `hash('evil')` to the meta tag, the optimizer should delete this, since `<script>evil</script>` is not part of the document.

Appending is *only* useful for when someone has external cross-origin scripts, in which case they need to add these CSP hashes manually - but that's likely to be an advanced use-case. Ideally, in the future, the optimizer should handle those as well.

Reply:
If `append=false` is default, then we need to disable the whole transformer by default. Otherwise, we might break existing pages with cross origin amp-script. If we disable by default, then the security advantage is not given either.

Ensuring security best practices so far hasn't been in the scope of AMP Optimizer.

Reply:
How about making it no-op by default if a `<meta name="amp-script-src" ...>` exists?

So we could have three "modes":

1.  Do nothing when CSP exists (default).
2.  Append when CSP exists.
3.  Replace when CSP exists.

Reply:
Went ahead and did this. `append` is now `mode`.

Reply:
perfect!